### PR TITLE
FilterHandler refactor

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Refactoring
- Documentation

### Description

_Please describe your pull request_

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [x] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [x] Reference relevant issue(s) and close them after merging
- [x] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by k-wall** [comment]
The PR is still work in progress.

**Comment by k-wall** [comment]
@robobario @SamBarker what do you think of this refactor?  hopefully the commit history should help you see how it ends up where it is.

**Comment by robobario** [comment]
Sorry, I didn't get to this today.

**Comment by SamBarker** [comment]
I'm sorry too. I thought I had cleared the backlog too.

**Comment by k-wall** [comment]
This PR is now to a state where I'm happy with it.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (robobario) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:288]: So is this triggering a flush after the current queue of work is complete?

**Comment by robobario** [comment]
The refactor looks good in general, I'm still reviewing, got caught up looking at [this bug](https://github.com/kroxylicious/kroxylicious/issues/536#issuecomment-1670978420)

About the change adding in chained ctx.flushes, what problem did that fix?

**Comment by k-wall** [comment]
> The refactor looks good in general, I'm still reviewing, got caught up looking at [this bug](https://github.com/kroxylicious/kroxylicious/issues/536#issuecomment-1670978420)
> 
> About the change adding in chained ctx.flushes, what problem did that fix?


I discovered a bug when I was trying to increase coverage.

File: api/kroxylicious-api/kroxylicious-api.iml
Comment (robobario) [inline api/kroxylicious-api/kroxylicious-api.iml:1]: The phantom iml. We should revisit if we can switch to a more typical idea gitignore now.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (robobario) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:117]: Opaque frames need to go on the queue too, right? If this filter is awaiting some async computation triggered by a decoded frame we don't want an OpaqueResponseFrame to be forwarded on to the client and responding out of order.

Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:117]: PR here https://github.com/k-wall/kroxylicious/pull/6

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:175]: Personally I find the logic easier to follow if I add, `defer` as an arg to `handleDeferredStage` and`handleDeferredReadCompletion`. If its false simply return the passed in future.

Passing `defer` seems like a bit of a smell, but we need to make consistent decisions between `handleDeferredStage` and `handleDeferredReadCompletion`so we need to decide once if we are defering or not.

That means this block ends up looking like 
```suggestion
boolean defer = !future.isDone();
        var filterHandled = handleDeferredStage(decodedFrame, future, defer)
                .thenApply(FilterHandler::validateFilterResultNonNull)
                .thenApply(fr -> handleRequestFilterResult(decodedFrame, promise, fr))
                .exceptionally(t -> handleFilteringException(t, decodedFrame));
        return handleDeferredWriteCompletion(filterHandled, defer)
                .thenApply(filterResult -> null);
```

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (SamBarker) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:149]: I'm not sure about the method name and I think that leads to it not having a single responsibility. 

Personally I think this is actually `dispatchDecodedResponse` and it potentially splits into `invokeFilterOnResponse` which covers L150 to L165. The result of that should be handed off to something else `forwardFilterResponse` covering the future management?  

I'm not wedded to any of the naming suggestions but I think this deserves to be broken down further.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (robobario) [inline kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:106]: I think I've found a bug in this and the equivalent write code, it doesn't queue properly. If you have multiple actions in the queue, the first future completion will call `readDecodedResponse` on all of the enqueued actions. This is visible when you set up the test with two deferred futures followed by a nondeferred future, the response is out of order because completing future1 triggers both actions.

Here's a fix, switching to thenCompose to chain the futures https://github.com/k-wall/kroxylicious/pull/7

Also I was wondering if we need an immediate `flush` after each deferred write finishes, if we wait for the whenComplete on the `writeFuture` then we may be delayed while other async work completes, where we were able to flush it to the broker straight away.

Comment (robobario) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:106]: Existing bug so we could leave it out of this PR

**Review by robobario — APPROVED** [review]
LGTM once opaque frame handling sorted. Thanks @k-wall

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (SamBarker) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:175]: I went a slightly different direction in https://github.com/k-wall/kroxylicious/pull/8

**Review by SamBarker — APPROVED** [review]
I ended up creating https://github.com/k-wall/kroxylicious/pull/8 as that expressed my previous thoughts more cohesively but I don't believe they are necessary for this to be merged.

Assuming the issue @robobario identified with the queuing of Opaque frames is fixed I'm happy to merge.

File: api/kroxylicious-api/kroxylicious-api.iml
Comment (k-wall) [reply api/kroxylicious-api/kroxylicious-api.iml]: removed .iml from the PR.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:117]: Great catch. I've added to this PR.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: The change looks good.  I'll keep it in this PR, but merged as separate commit.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:180]: I like the idea of pushing the responsibility to the SafeInvoker.  Let's do it as a PR after this one land.

Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:180]: https://github.com/kroxylicious/kroxylicious/issues/558

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:163]: https://github.com/k-wall/kroxylicious/pull/8 looked good. I've incorporated into this PR.

File: kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java
Comment (k-wall) [reply kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: I liked the idea, it is now incorporated into the PR.  I kept it a separate commit.

**Pull request merged**
